setup.py

```diff
@@ -23,8 +25,87 @@
     sys.exit(1)
 
-sys.path.insert(0, os.path.abspath('lib'))
-from ansible.release import __version__, __author__
+
+def find_package_info(*file_paths):
+    # Open in Latin-1 so that we avoid encoding errors.
+    # Use codecs.open for Python 2 compatibility
+    try:
+        f = codecs.open(os.path.join(*file_paths), 'r', 'latin1')
+        info_file = f.read()
+        f.close()
+    except:
+        raise RuntimeError("Unable to find package info.")
+
+    # The version line must have the form
+    # __version__ = 'ver'
+    version_match = re.search(r"^__version__ = ['\"]([^'\"]*)['\"]",
+                              info_file, re.M)
+    author_match = re.search(r"^__author__ = ['\"]([^'\"]*)['\"]",
+                             info_file, re.M)
+
+    if version_match and author_match:
+        return version_match.group(1), author_match.group(1)
+    raise RuntimeError("Unable to find package info.")
+
+
+def _validate_install_ansible_base():
+    """Validate that we can install ansible-base. Currently this only
+    cares about upgrading to ansible-base from ansible<2.10
+    """
+    if os.getenv('ANSIBLE_SKIP_CONFLICT_CHECK', '') not in ('', 0):
+        return
+
+    # Save these for later restoring things to pre invocation
+    sys_modules = sys.modules.copy()
+    sys_modules_keys = set(sys_modules)
+
+    # Make sure `lib` isn't in `sys.path` that could confuse this
+    sys_path = sys.path[:]
+    abspath = os.path.abspath
+    sys.path[:] = [p for p in sys.path if abspath(p) != abspath('lib')]
+
+    try:
+        from ansible.release import __version__
+    except ImportError:
+        pass
+    else:
+        version_tuple = tuple(int(v) for v in __version__.split('.')[:2])
+        if version_tuple < (2, 10):
+            stars = '*' * 76
+            raise RuntimeError(
+                '''
+
+    %s
+
+    Cannot install ansible-base with a pre-existing ansible==%s
+    installation.
+
+    Installing ansible-base with ansible-2.9 or older currently installed with
+    pip is known to cause problems. Please uninstall ansible and install the new
+    version:
+
+        pip uninstall ansible
+        pip install ansible-base
+
+    If you want to skip the conflict checks and manually resolve any issues
+    afterwards, set the ANSIBLE_SKIP_CONFLICT_CHECK environment variable:
+
+        ANSIBLE_SKIP_CONFLICT_CHECK=1 pip install ansible-base
+
+    %s
+                ''' % (stars, __version__, stars)
+                # '\n\n%s\n\nCannot upgrade ansible==%s to ansible-base. You must first uninstall\n'
+                # 'ansible before installing ansible-base. This is only required when upgrading\n'
+                # 'from ansible<2.10 to ansible-base.\n\n%s\n' % (stars, __version__, stars)
+            )
+    finally:
+        sys.path[:] = sys_path
+        for key in sys_modules_keys.symmetric_difference(sys.modules):
+            sys.modules.pop(key, None)
+        sys.modules.update(sys_modules)
```

Comment:
What is this supposed to achieve? Modules are cached on import, they'll be picked up from that cache regardless of the pointers being listed in `sys.modules`.

Reply:
This is just setting things back to the way they were, the `.update()` is likely pointless as it stands, but it was to safe guard some future state where the act of importing a module, has the side effect of deleting some other module from `sys.modules`.